### PR TITLE
Implement LRA Theory Solver to be used to handle relational assumptions in new assumptions

### PR DESCRIPTION
**Pull request by TiloRC**
Continued in https://github.com/sympy/sympy/pull/25544

Adds a linear real arithmetic theory solver to logic/algorithms. It has the ability to check if conjunctions of equalities and strict and non-strict inequalities are satisfiable and is designed to work with encoded cnf objects.  

#25544 integrates the code from this pull request with dpll2.py.

I haven't done any tests, but I suspect the speed bottle neck is how it processes inequalities in `from_encoded_cnf`. Ideally the inequalities would come preprocessed. 

#### Statistics 

To better understand what was slowing down, I measured how long each function in the LRASolver was taking to run in random tests. `from_encoded_cnf()` does preprocessing once for each problem, `assert_enc_boundry()` activates each constraint in the problem, and `check()` is where the main algorithm runs to check if the problem is satisfiable. 

EDIT: Since running the tests, I have made improvements to `from_encoded_cnf()` which speed it up dramatically. 

##### Problems of Varying Sizes

|            | `from_encoded_cnf()` | `assert_enc_boundry()` | `check()` | Total Time (s) | % sat |
|---------------------|--------------------|----------------------|-------|-------|-------|
| 2 var, 4 const  | 12.4               | 0.129                | 0.922 | 13.5  | 51    |
| 4 var, 8 const  | 43.0               | 0.265                | 7.00  | 50.2  | 41    |
| 8 var, 16 const  | 150                | 0.555                | 64.7  | 214.77| 28    |
| 16 var, 32 const | 584 | 1.28 | 1000 | 1590 | 14

##### Integer vs Rational Coefficients

|            | `from_encoded_cnf()` | `assert_enc_boundry()` | `check()` | Total Time (s) | % sat |
|---------------------|--------------------|----------------------|-------|-------|-------|
| integer coeff | 43.0               | 0.265                | 7.00  | 50.2  | 41    |
| rational coeff  | 60.6                | 0.264                | 8.80  | 69.6 | 41    |

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by smichr** [comment]
Do you have (or have you seen) a nice example of where the solution from a system of strict inequalities is significantly different from one where the inequalities are not? And by significantly different I mean not including the case where `a<b` is false but `a<=b` is true.

File: sympy/solvers/tests/test_inequalities.py [hunk 1]
```diff
@@ -486,3 +490,127 @@ def test__pt():
...
+# lra = LRASolver(A, [s1, s2], [x, y])
+# lra.assert_con(x <= -4)
+# lra.assert_con(x >= -8)
+# lra.assert_con(s1 <= 1)
+# lra.assert_con(s2 >= -3)
+# lra.check()
+
+
+
+
```
Comment (smichr) [inline sympy/solvers/tests/test_inequalities.py:616]:
```suggestion
```

**Comment by TiloRC** [comment]
> Do you have (or have you seen) a nice example of where the solution from a system of strict inequalities is significantly different from one where the inequalities are not?

I'm not exactly sure what you mean by this. Does this count as significantly different?

Has no solutions:

x+y>0,  x<0,  y<0

Has a solution at x=0, y=0:

x+y>=0,  x=<0,  y=<0

Why do you ask? Is it because you want to try to implement the find feasible function using the code from the linear programing PR? I know I said that wasn't possible, but it might be. It's just that the approach the paper I'm using to implement this PR won't work.

Basically what this paper does is that strict inequalities like x < 0 become x + delta <= 0 where delta is an infinitely small quantity. The problem with doing this in the in the linear programing PR is that I don't think multiplication/division between terms like  x + delta is well defined. 

If you're interested in solutions for linear programing problems with strict inequalities, I don't think this is possible. For example, it's not possible to get a maximum value for x subject to x < 0.

**Comment by oscarbenjamin** [comment]
> Has no solutions:
> 
> x+y>0, x<0, y<0
> 
> Has a solution at x=0, y=0:
> 
> x+y>=0, x=<0, y=<0

One way to approach this is to rewrite strict inequalities like:
```
x > 0    -->   x >= 0, x != 0
x + y > 1   -->   z = x + y - 1, z >= 0, z != 0
```
The unequalities would not be included in the linear programming problem but they could be checked afterwards: if the only feasible value of z was zero then that can be rejected. Otherwise the condition `z != 0` can give us a symbolic condition on the variables (`x + y - 1 != 0`) that can be used to constrain any conclusions generated using `z >= 0`.

**Comment by TiloRC** [comment]
It seems like there are probably a lot of ways to go about dealing with strict inequalities. I think I'm probably going to go with what the paper I'm implementing does. This paper is used by Z3 so the approach must be pretty fast.

**Comment by smichr** [comment]
> Why do you ask?

It just seems to me that if you have a solution for non-strict inequalities, if you change them all to strict inequalities the solution will almost be the same -- but you will be located eps away from the solution. So if an inequality was y < 3 and the solution contained y = 3 then you would have to make the solution 3 - var('eps', positive=True). But as someone says [here](https://math.stackexchange.com/questions/2673917/can-a-linear-program-have-strict-inequalities) a minimization not on the boundary is strange -- you have to tell what values *can* be part of the solution (hence closed space).

Perhaps what is different is that in checking feasibility, one wants to know if you *can* have a solution somewhere, not necessarily max or min. But that seems similar to the max/min problem since if you can have a max or min, the system is feasible. It will be interesting to compare the approaches when you are done.

**Comment by TiloRC** [comment]
> Perhaps what is different is that in checking feasibility, one wants to know if you can have a solution somewhere, not necessarily max or min.

This is my understanding. 

> It will be interesting to compare the approaches when you are done.

Cool! I presume you will implement something similar then?

**Comment by TiloRC** [comment]
I'm not sure where the best place to put the code in this PR is. Maybe a better home for it would be sympy.logic.inference.

**Comment by smichr** [comment]
> presume you will implement something similar then?

no -- just the feasibility step of `lp` (or whatever it will be called). I am thinking of sticking with the scipy convention and making an optimize module and call the function `linprog`. I'm not sure if that will be confusing, however, since our inputs are different than theirs.

**Comment by smichr** [comment]
> Maybe a better home for it would be sympy.logic.inference

hmm...Instead of an optimize maybe a module for inequality_systems could be added. Then `lp` and `is_feasible` could be there.

**Comment by TiloRC** [comment]
> Then `lp` and `is_feasible` could be there.

It might make sense to get rid of `is_feasible` because [`satisfiable`](https://docs.sympy.org/latest/modules/logic.html#sympy.logic.inference.satisfiable) will be able to do everything `is_feasible` can do and more.

**Comment by smichr** [comment]
> It might make sense to get rid of `is_feasible`

I meant whatever function you are working on.

**Comment by TiloRC** [comment]
> no -- just the feasibility step of lp (or whatever it will be called).

I think it would make most sense to raise an exception if the input contains a strict inequality in the context of linear programing. That's what other software does.

**Comment by TiloRC** [comment]
> I meant whatever function you are working on.

`satisfiable` is an existing function. My idea is rather than making a new function just make `satisfiable` handle inequalities.

File: sympy/solvers/inequalities.py [hunk 2]
```diff
@@ -982,3 +988,338 @@ def reduce_inequalities(inequalities, symbols=[]):
...
+        iteration = 0
+        while True:
+            iteration += 1; _debug_internal_state_printer1(iteration, M, basic, self.all_var)
+
+            # nonbasic variables always must be within bounds
+            assert all(((self.assign[nb] >= self.lower[nb]) == True) and ((self.assign[nb] <= self.upper[nb]) == True) for nb in nonbasic)
+
+            # assignments for x must always satisfy Ax = 0
+            X = Matrix([self.assign[v] for v in self.col_index])
+            assert all(abs(val) < 10**(-10) for val in M*X) # TODO: Maybe return None if this cond can't be met?
```
Comment (TiloRC) [inline sympy/solvers/inequalities.py:1182]: Maybe we can have a variable that's set when the tests are being run so that these assert statements are only run during the tests. Maybe pytest has a way to this.

Comment (asmeurer) [reply sympy/solvers/inequalities.py:1182]: @oscarbenjamin any thoughts on how this could be done? Is there a variable that gets set when the tests are run?

File: sympy/solvers/inequalities.py [hunk 3]
```diff
@@ -982,3 +988,338 @@ def reduce_inequalities(inequalities, symbols=[]):
...
+
+            cand = [b for b in basic
+             if self.assign[b] < self.lower[b]
+             or self.assign[b] > self.upper[b]]
+
+            if len(cand) == 0:
+                return "SAT", self.assign
```
Comment (TiloRC) [inline sympy/solvers/inequalities.py:1189]: Use True/False for "SAT"/"UNSAT"

File: sympy/solvers/inequalities.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/solvers/inequalities.py:1182]: Why not have a parameter like `check=False` and then call the function with check=True during the test code? Then it is easier to enable checking in any situation rather than just the test suite.

The kind of thing that I would potentially want an assert like this for is to pick up on problems that might arise in cases that are *not* tested e.g. because user code does something unanticipated. If the asserts only run during testing then they can just be part of the test suite.

**Comment by TiloRC** [comment]
This comment is mostly for my own future reference.
 
I'm not totally sure what to do with equality. Here are some options: 

### in a preprocessing step split each equality into two inequalities

I can't do this because equality is not necessarily equivalent to two inequalities. Inequalities imply their arguments are real numbers whereas equalities do not. 

### Add clauses that convey that Q.eq(x, y) implies x <= y and x>=y

This would slow things down. Also possibly a bit involved.  

### Have LRASolver deal with equality internally 

The difficulty here is that generating conflict clauses becomes very difficult as conflict clauses are made up of only inequalities. Perhaps I can keep track of whether the current boundaries for each variable are the result of an equality or not. 

I think the LRASolver should probably deal with equality internally. This will be a bit messy, but I think it's the best option.

**Comment by TiloRC** [comment]
Things I want to get done by the meeting tomorrow: 
- equality should be handled 
- conflict clauses should be literals
- clean up all the junk/old code that's no longer needed 
- if I have time, start thinking about back tracking

File: sympy/logic/tests/test_lra_theory.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    # lra.assert_con(y <= 0)
+    # lra.assert_con(s1 >= 2)
+    # assert lra.check() == ('UNSAT', {s1 >= 2, x <= 0, y <= 0})
+    #
+    #
+    # # test potential edge case
+    # r1 = x <= 1
+    # r2 = -x <= -5
+    # lra = LRASolver(Matrix(), [], [x])
+    # lra.assert_con(x <= 1)
```
Comment (TiloRC) [inline sympy/logic/tests/test_lra_theory.py:195]: This code uses depreciated functions. I'm not sure if I should convert it into code that uses the new functions or just get rid of it. I feel like it's kind of pointless because test_random_problems() should be able to handle all of the cases here.

File: sympy/logic/tests/test_lra_theory.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    assert len(lra.A) == 0
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.ge(x, 0))) +1) == ("OK", None)
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.ge(x, 1))) +1) == ("OK", None)
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.le(x, -1))) + 1) == ('UNSAT', {-(enc.symbols.index((Q.le(x, -1))) + 1), -(enc.symbols.index((Q.ge(x, 1))) +1)})
+    # assert lra.assert_con(Q.ge(x, 0)) == ("OK", None)
+    # assert lra.assert_con(Q.ge(x, -1)) == ('SAT', {x: 0, y: 0})
+    # assert lra.assert_con(Q.le(x, -1)) == ('UNSAT', {x <= -1, x >= 0})
+    #
+    # m = Matrix()
+    # lra = LRASolver(m, [], [x, y])
```
Comment (TiloRC) [inline sympy/logic/tests/test_lra_theory.py:143]: This commented out code code uses depreciated functions. I'm not sure if I should convert it into code that uses the new functions or just get rid of it. I feel like it's kind of pointless because test_random_problems() should be able to handle all of the cases here.

**Comment by TiloRC** [comment]
https://github.com/joerick/pyinstrument

File: sympy/logic/algorithms/lra_theory.py [hunk 6]
```diff
@@ -0,0 +1,667 @@
...
+
+    if prop.function == Q.eq:
+        return Q.eq(var, const)
+    if prop.function in [Q.gt, Q.lt]:
+        return Q.lt(var, const)
+    else:
+        return Q.le(var, const)
+
+
+class Boundry:
```
Comment (asmeurer) [inline sympy/logic/algorithms/lra_theory.py:68]: This could just use a dataclass https://docs.python.org/3/library/dataclasses.html

**Comment by TiloRC** [comment]
Continued in #25544. Thought it was confusing to have PR with overlapping code.
